### PR TITLE
Improve formatting features in the editor

### PR DESCRIPTION
**Pull request by alexanderstephan**
**Description**: 

This PR aims to improve, the usability of the editor formatting features.

- It fixes https://github.com/vector-im/element-web/issues/19501  by adding a short cut for inserting links. I used `CTRL` + `ALT` + `K` to stay consistent with other tools that usually use `CTRL` + `K` (which unfortunately is already used by Element to search for rooms). I did not choose `CTRL` + `SHIFT` + `K` , as this is commonly used for developer tools by browsers. (This has been changed to `CTRL` + `SHIFT` + `L` for now, see comments below, the link formatting can now also be toggled).

- It adds `CTRL` + `E` as a shortcut for formatting code. I chose this shortcut, because for example GitHub also uses it and it seems to pretty common. I am not aware of any conflicts.

- Code formatting can now also be toggled, like it already did for inline formatting such as `bold`. It now works for inline and even multiline formatting. Right now, it would just put more an more backticks around the selection.

- **For all formatting actions that have shortcuts:** When no text is selected by the user and a format action is toggled via a shortcut, the word the caret is currently at gets automatically selected and formatted. Afterwards the caret will be adjusted to to match the previous position (relative to the current word). When the caret is at the position of the formatting string itself, it snaps to the word, the formatting belongs to (can be seen in the last gif). This behavior is similar to the GitHub editor.

- Fixes a bug where formatting white space would cause weird behavior as the trim of the selection did not work properly. Right now, it is not possible to format whitespace. This is also consistent how it works in Telegram for example.

**Demo:**

Formatting a link without previous selection with shortcut:
Links

Toggling multi line code formatting:
Multiline

How a typical work flow could look like: 
Peek 2021-11-09 16-28

Right punctuation is included (!), but is worth to discuss what's better here. For example when formatting a `myObj.x` only `myObj` would get selected when the caret e.g. is before `y` which might not to be expected.

**Signed-off-by:** Alexander Stephan alexander.stephan@tum.de

Fixes https://github.com/vector-im/element-web/issues/19501

---
Here's what your changelog entry will look like:

## ✨ Features
 * Improve formatting features in the editor ([\#7104](https://github.com/matrix-org/matrix-react-sdk/pull/7104)). Fixes vector-im/element-web#19501. Contributed by @alexanderstephan.

Preview: https://pr7104--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/editor/operations.ts
Comment (t3chguy) [inline src/editor/operations.ts:41]:
```suggestion
    if (range.length === 0) {
```

**Review by t3chguy — COMMENTED** [review]
Implementation looks sane and the additional tests are great. This PR solves a lot of edge cases and that is great, one small issue with the designated shortcut seemingly not working on macOS (tested on Chrome browser on the Netlify build)

File: src/KeyBindingsDefaults.ts
Comment (alexanderstephan) [reply src/KeyBindingsDefaults.ts:94]: I just tested on macOS and also found that it does not work on Firefox either.
I think <kbd>Ctrl</kbd> + <kbd>Alt</kbd> + <kbd>K</kbd> does not work since it is used for the Δ character on macOS. I am not quite sure why<kbd>Cmd</kbd> + <kbd>Alt</kbd> + <kbd>K</kbd> does not work, but it also does not seem to be a good choice, because under macOS it opens dev tools on Firefox. I think it would be best to switch between <kbd>Shift</kbd> (macOS) and <kbd>Alt</kbd> (All others).  This way the shortcuts also work as expected.

File: src/KeyBindingsDefaults.ts
Comment (t3chguy) [inline src/KeyBindingsDefaults.ts:93]: this seems to say `Shift + K` on macOS which seems to conflict with typing an uppercase K? Presumably the ctrlOrCmd should stick around?

Comment (alexanderstephan) [reply src/KeyBindingsDefaults.ts]: Yes, you are totally right, I accidentally deleted it ... should be fixed now.

**Comment by t3chguy** [comment]
I can confirm the shortcut works, would it be possible to have the shortcut toggle the link? Right now it just infinitely nests which isn't the most useful behaviour

**Comment by alexanderstephan** [comment]
I tried to mimic the behavior from GitHub here, but I agree, this is not really too covenient. I can't think of any real good reason why it shouldn't get toggledas well. I will add this!
Also, I think we could use <kbd>Ctrl/Cmd</kbd> +  <kbd>Shift </kbd> +  <kbd>L </kbd> for the shortcut. It should not conflict and be consistent across platforms. Also L being a mnemonic for link. Right now, it might mess up muscle memory when switching devices often. What do you think?

**Comment by alexanderstephan** [comment]
@janogarcia @t3chguy Any updates or suggestions on this?

File: src/editor/operations.ts
Comment (UwUnyaa) [inline src/editor/operations.ts:128]: Is this list of characters sufficient? I don't think punctuation and parentheses should be considered parts of the word, but the specifics are up to discussion.

Comment (alexanderstephan) [reply src/editor/operations.ts:128]: I mean word is used a bit freely here, but I don't really have any better name right now.
It basically boils down to where you want to auto-selection for formatting to stop.
I think one usually wants to include parentheses and punctuation, but I actually excluded these previously, so I am also not 100% sure about this design-wise. Technically, the `+` could be also removed to match GitHub's behavior for example.

File: src/editor/operations.ts
Comment (alexanderstephan) [reply src/editor/operations.ts:95]: Yes, I modified the signature in the end. Good catch! :+1:

**Comment by t3chguy** [comment]
Annoyingly Cmd Shift L seems to conflict with Bitwarden :')

**Review by t3chguy — APPROVED** [review]
Looks sane to me otherwise

**Comment by alexanderstephan** [comment]
Sad, I also use Bitwarden. But is it really a problem? I can't see a use case where one would need auto-fill when having the messaging field in focus. Otherwise I don't notice any problems or conflicts as I also have the browser plugin installed.

**Comment by t3chguy** [comment]
No that's absolutely fine, just meant I didn't test the macos shortcut

**Comment by alexanderstephan** [comment]
@t3chguy Sorry, I don't want to annoy you again, but we are still waiting for the review of the designer, right? Some merge conflicts already arised and I wondered what could be done to get this merged soon, since it has been open for quite a while now.

**Comment by t3chguy** [comment]
> @t3chguy Sorry, I don't want to annoy you again, but we are still waiting for the review of the designer, right?

Yes

image

**Review by janogarcia — APPROVED** [review]
**Code formatting**

`Ctrl/⌘ + E` is working great for inline and multiline code snippets! 👌

**Formatting the previous word next to caret**

That's a really nice addition too, and it's working smoothly. 👏

**Link shortcut**

It's a shame we can't use the more "standard" `⌘K` shortcut, or `⌘⇧K` as you originally suggested, but yeah `⌘⇧L` (Mac) / `Ctrl + Shift + L` (Win) seems to be the second best option. 👍

Thanks for implementing all of these valuable improvements to the composer.

**Comment by janogarcia** [comment]
@alexanderstephan @t3chguy Just realized that the keyboard shortcut help dialog hasn't been updated to include the new ones. That would need to be done as a fast follow item.

**Comment by alexanderstephan** [comment]
@t3chguy @janogarcia Great catch, I did not know this overview existed. I will fix this as soon as possible. I will also do the remaining merge.

**Comment by alexanderstephan** [comment]
Just to be sure: I am not so familiar with the terminology, do you mean submitting a follow up PR or fixing it in this PR?
As I have to merge anyways, I would like to do it in this PR.

**Comment by janogarcia** [comment]
@alexanderstephan You're very welcome to do it in this PR after merging!

What I meant was _"I don't want to block this PR anymore for this reason. Let's (Michale or me) take care of updating the keyboard shortcut dialog after Alexander has had the chance to merge this PR"_. 😊

**Comment by t3chguy** [comment]
I'll add it to this PR

**Comment by alexanderstephan** [comment]
@t3chguy Thanks so much! :+1:  I would like to add a german translation. Is this a good idea to do it here, or should it be done via Weblate later?

**Comment by t3chguy** [comment]
Must be done via Weblate later, otherwise it'll cause conflicts with Weblate's working copy of the i18n files

**Comment by alexanderstephan** [comment]
Ok, then I'll add it on Weblate, once it's possible. :+1:

**Comment by Garfield96** [comment]
Thank you @alexanderstephan, this is a really useful addition.
I hope this feature will be included in the next release.

**Review by andybalaam — APPROVED** [review]
Really improves the code while adding useful new features, thank you!

**Pull request merged**
